File: client/src/lib/reportLayouts.ts
import type { ReportType } from "../types/report";

export type PreviewSection =
  | "collaborators"
  | "developers"
  | "reviewers"
  | "testers"
  | "requesters"
  | "projects"
  | "cards"
  | "sla_dev"
  | "sla_cards"
  | "bottlenecks"
  | "alerts"
  | "team_summary"
  | "flow"
  | "priority"
  | "dora"
  | "quality_gates"
  | "discipline"
  | "individual"
  | "role_metrics"
  | "fibonacci"
  | "trends"
  | "risk"
  | "analysis_workflow"
  | "antifraud";

export const REPORT_PREVIEW_SECTIONS: Record<ReportType, PreviewSection[]> = {
  general: [
    "team_summary",
    "flow",
    "priority",
    "dora",
    "quality_gates",
    "discipline",
    "analysis_workflow",
    "antifraud",
    "fibonacci",
    "collaborators",
    "developers",
    "reviewers",
    "testers",
    "requesters",
    "projects",
    "sla_dev",
    "sla_cards",
    "bottlenecks",
    "alerts",
  ],
  individual: ["individual", "role_metrics", "collaborators"],
  developers: ["role_metrics", "developers", "flow", "fibonacci", "sla_dev", "quality_gates"],
  requesters: ["role_metrics", "requesters", "projects"],
  testers: ["role_metrics", "testers", "quality_gates"],
  management: [
    "team_summary",
    "flow",
    "priority",
    "dora",
    "discipline",
    "quality_gates",
    "analysis_workflow",
    "antifraud",
    "projects",
    "sla_dev",
    "bottlenecks",
    "alerts",
    "trends",
    "risk",
  ],
  specific_metrics: ["team_summary", "sla_dev", "sla_cards", "bottlenecks", "flow", "priority", "dora", "antifraud"],
};

export const METRIC_KEY_SECTIONS: Record<string, PreviewSection[]> = {
  team_summary: ["team_summary"],
  flow: ["flow"],
  developers: ["role_metrics", "developers", "fibonacci", "flow", "quality_gates", "sla_dev"],
  testers: ["role_metrics", "testers", "quality_gates"],
  requesters: ["role_metrics", "requesters", "projects"],
  sla: ["sla_dev", "sla_cards", "alerts"],
  bottlenecks: ["bottlenecks"],
  card_dossier: [],
  priority: ["priority"],
  dora: ["dora"],
  quality_gates: ["quality_gates"],
  discipline: ["discipline"],
  analysis_workflow: ["analysis_workflow"],
  antifraud: ["antifraud"],
};

export function allowedSectionsForReport(
  reportType: ReportType,
  metricKeys?: string[],
): Set<PreviewSection> {
  if (reportType === "specific_metrics" && metricKeys?.length) {
    const allowed = new Set<PreviewSection>();
    for (const key of metricKeys) {
      for (const section of METRIC_KEY_SECTIONS[key] ?? []) {
        allowed.add(section);
      }
    }
    return allowed;
  }
  return new Set(REPORT_PREVIEW_SECTIONS[reportType] ?? REPORT_PREVIEW_SECTIONS.general);
}

export const TAB_DESCRIPTIONS: Record<ReportType, string> = {
  general: "Visao completa do time: entregas, fluxo, qualidade, pessoas e dossie de cards.",
  individual: "Perfil consolidado de um colaborador e cards em que atuou no periodo.",
  developers: "Entregas, pontos Fibonacci, tempo de atuacao e qualidade por desenvolvedor.",
  requesters: "Demanda criada, planejamento e entregas por solicitante.",
  testers: "Testes, primeira passagem, problemas evitados e retestes.",
  management: "Indicadores para gestao: fluxo, SLA, risco, DORA, analises e tendencias.",
  specific_metrics: "Somente as metricas selecionadas abaixo.",
};

File: trello_metrics/reports/report_layouts.py
from __future__ import annotations

from typing import Any

# Secoes que o PDF/HTML podem renderizar por tipo de relatorio.
REPORT_LAYOUTS: dict[str, dict[str, Any]] = {
    "general": {
        "title_prefix": "Relatorio geral",
        "sections": [
            "overview",
            "team_summary",
            "metric_guide",
            "ai",
            "flow",
            "fibonacci",
            "collaborators",
            "developers",
            "reviewers",
            "formal_reviewers",
            "testers",
            "requesters",
            "projects",
            "sla",
            "bottlenecks",
            "trends",
            "quality_gates",
            "movements",
            "analysis_workflow",
            "antifraud",
            "dossier",
        ],
    },
    "individual": {
        "title_prefix": "Relatorio individual",
        "sections": ["overview", "metric_guide", "ai", "individual_summary", "dossier"],
    },
    "developers": {
        "title_prefix": "Relatorio de desenvolvedores",
        "sections": [
            "overview",
            "metric_guide",
            "ai",
            "role_summary",
            "flow",
            "fibonacci",
            "developers",
            "sla",
            "quality_gates",
            "dossier",
        ],
    },
    "requesters": {
        "title_prefix": "Relatorio de solicitantes",
        "sections": ["overview", "metric_guide", "ai", "role_summary", "requesters", "projects", "dossier"],
    },
    "testers": {
        "title_prefix": "Relatorio de testers",
        "sections": ["overview", "metric_guide", "ai", "role_summary", "testers", "quality_gates", "dossier"],
    },
    "reviewers": {
        "title_prefix": "Relatorio de revisao em par",
        "sections": ["overview", "metric_guide", "ai", "role_summary", "reviewers", "quality_gates", "dossier"],
    },
    "formal_reviewers": {
        "title_prefix": "Relatorio de revisores",
        "sections": [
            "overview",
            "metric_guide",
            "ai",
            "role_summary",
            "formal_reviewers",
            "quality_gates",
            "dossier",
        ],
    },
    "management": {
        "title_prefix": "Relatorio de gestao",
        "sections": [
            "overview",
            "team_summary",
            "metric_guide",
            "ai",
            "flow",
            "risk",
            "priority",
            "discipline",
            "projects",
            "sla",
            "bottlenecks",
            "trends",
            "quality_gates",
            "analysis_workflow",
            "antifraud",
        ],
    },
    "specific_metrics": {
        "title_prefix": "Relatorio de metricas especificas",
        "sections": ["overview", "metric_guide", "role_summary", "ai", "dynamic"],
    },
}


def report_layout(report_type: str | None) -> dict[str, Any]:
    return REPORT_LAYOUTS.get(report_type or "general", REPORT_LAYOUTS["general"])


def allows_section(report_type: str | None, section: str) -> bool:
    layout = report_layout(report_type)
    sections = layout.get("sections") or []
    if "dynamic" in sections:
        return section in sections or section not in _FIXED_SECTIONS
    return section in sections


_FIXED_SECTIONS = {
    "overview",
    "team_summary",
    "individual_summary",
    "role_summary",
    "metric_guide",
    "ai",
    "flow",
    "fibonacci",
    "collaborators",
    "developers",
    "reviewers",
    "formal_reviewers",
    "testers",
    "requesters",
    "projects",
    "sla",
    "bottlenecks",
    "trends",
    "quality_gates",
    "movements",
    "dossier",
    "risk",
    "priority",
    "discipline",
    "analysis_workflow",
    "antifraud",
}
